server/src/main/java/org/elasticsearch/common/time/DateUtilsRounding.java

```diff
@@ -92,8 +92,27 @@ static long utcMillisAtStartOfYear(final int year) {
         return (year * 365L + (leapYears - DAYS_0000_TO_1970)) * MILLIS_PER_DAY; // millis per day
     }
 
-    private static boolean isLeapYear(final int year) {
-        return ((year & 3) == 0) && ((year % 100) != 0 || (year % 400) == 0);
+    static boolean isLeapYear(final int year) {
+        // Joda had
+        // return ((year & 3) == 0) && ((year % 100) != 0 || (year % 400) == 0);
+        // But we've replaced that with this:
+        if ((year & 3) != 0) {
+            return false;
+        }
+        if (year % 100 != 0) {
+            return true;
+        }
+        return ((year / 100) & 3) == 0;
```

Comment:
Something I was thinking about - do we see the difference because of the change from `year % 400` to `(year / 100) & 3`, or because of unrolling the short circuit logic?  And if it's only the former, does it make more sense to leave this as a one-liner and let the optimizer short circuit it as necessary?  I'm more asking out of curiosity about how the optimizer works than out of a need to see this changed.

Reply:
I think the short circuiting is actually the same here. At least, that is how I read the Java. I believe the speed is entirely around the change from `(year % 400) == 0` to `((year / 100) & 3) == 0`. We certainly could keep it as a one liner, but I think it as harder to read that way. Certainly harder for me to track down the disassembly chunks because the disassembler sticks line numbers in there. Not that you can truly believe them though.